src/AppInstallerCLICore/Commands/DscSourceResource.cpp

```diff
@@ -0,0 +1,429 @@
+    {
+        if (auto json = GetJsonFromInput(context))
+        {
+            SourceFunctionData data{ context, json };
+
+            data.Get();
+
+            WriteJsonOutputLine(context, data.Output.ToJson());
+        }
+    }
+
+    void DscSourceResource::ResourceFunctionSet(Execution::Context& context) const
+    {
+        if (auto json = GetJsonFromInput(context))
+        {
+            SourceFunctionData data{ context, json };
+
+            data.Get();
+
+            // Capture the diff before updating the output
+            auto diff = data.DiffJson();
+
+            if (!data.Test())
+            {
+                if (data.Input.ShouldExist())
+                {
+                    if (data.Output.Exist().value())
+                    {
+                        AICLI_LOG(CLI, Info, << "Replacing source with new information");
+                        data.Replace();
+                    }
+                    else
+                    {
+                        AICLI_LOG(CLI, Info, << "Adding source as it was not found");
+                        data.Add();
+                    }
+                }
+                else
+                {
+                    AICLI_LOG(CLI, Info, << "Removing source as desired");
+                    data.Remove();
+                }
+
+                if (data.SubContext->IsTerminated())
+                {
+                    data.ParentContext.Terminate(data.SubContext->GetTerminationHR());
+                    return;
+                }
+
+                data.Reset();
+                data.Get();
+            }
+
+            WriteJsonOutputLine(context, data.Output.ToJson());
+            WriteJsonOutputLine(context, diff);
+        }
+    }
+
+    void DscSourceResource::ResourceFunctionTest(Execution::Context& context) const
+    {
+        if (auto json = GetJsonFromInput(context))
+        {
+            SourceFunctionData data{ context, json };
+
+            data.Get();
+            data.Output.InDesiredState(data.Test());
+
+            WriteJsonOutputLine(context, data.Output.ToJson());
+            WriteJsonOutputLine(context, data.DiffJson());
+        }
+    }
+
+    void DscSourceResource::ResourceFunctionExport(Execution::Context& context) const
+    {
+        auto currentSources = Repository::Source::GetCurrentSources();
+
+        for (auto const& source : currentSources)
+        {
+            SourceResourceObject output;
+            output.SourceName(source.Name);
+            output.Argument(source.Arg);
+            output.Type(source.Type);
+            output.TrustLevel(TrustLevelStringFromFlags(source.TrustLevel));
+            output.Explicit(source.Explicit);
+            WriteJsonOutputLine(context, output.ToJson());
+        }
+    }
+
+    void DscSourceResource::ResourceFunctionSchema(Execution::Context& context) const
+    {
+        WriteJsonOutputLine(context, SourceResourceObject::Schema(ResourceType()));

```

Comment:
Is this likely identical in all resources? Should we make a default implementation in parent?

Reply:
No, as each resource has its own "Object" type.